src/metabase/upload.clj

```diff
@@ -533,59 +539,75 @@
-        normalized-header+header (into []
-                                       (map (fn [col]
-                                              [(normalize-column-name col) col]))
-                                       header)
-        normalized-header        (map first normalized-header+header)
-        ;; check for duplicates
-        _ (when (some #(< 1 %) (vals (frequencies normalized-header)))
-            (throw (ex-info (tru "The CSV file contains duplicate column names.")
-                            {:status-code 422})))
+        normalized-header+header (->> header
+                                      (map (fn [col]
+                                             [(normalize-column-name col) col]))
+                                      remove-auto-pk-cols)
+        normalized-header (map first normalized-header+header)
         [extra missing _both] (data/diff (set normalized-header)
-                                         (set normalized-field-names))]
-
+                                         ;; ignore the auto-pk field, it will be filled automatically
+                                         (disj (set normalized-field-names) auto-pk-column-name))]
+    ;; check for duplicates
+    (when (some #(< 1 %) (vals (frequencies normalized-header)))
+      (throw (ex-info (tru "The CSV file contains duplicate column names.")
+                      {:status-code 422})))
     (if (and (empty? extra) (empty? missing))
       {:parse-rows (fn [rows]
-                     (let [rows' rows ;; TODO: remove auto-pk columns here
+                     (let [rows' (cond->> rows
+                                   (seq auto-pk-col-indices)
+                                   (map remove-auto-pk-cols))
                            col-upload-types (map (comp base-type->upload-type :base_type fields-by-normed-name)
                                                  normalized-header)]
                        (parse-rows* col-upload-types rows')))
-       :table-col-names (map (comp :name fields-by-normed-name) normalized-header)}
+       :table-col-names (map (comp :name fields-by-normed-name) normalized-header)
+       :table-has-auto-pk? (contains? fields-by-normed-name auto-pk-column-name)}
       (throw (ex-info (tru "The schema of the CSV file does not match the schema of the table.")
                       {:status-code 422})))))
 
 (defn- append-csv!*
   [database table file]
   (with-open [reader (bom/bom-reader file)]
     (let [[header & rows] (csv/read-csv reader)
-          {:keys [parse-rows table-col-names]} (check-schema table header)]
+          {:keys [parse-rows table-col-names table-has-auto-pk?]} (check-schema table header)
+          driver (driver.u/database->driver database)
+          auto-pk-type (driver/upload-type->database-type driver ::auto-incrementing-int-pk)
+          create-auto-pk? (not table-has-auto-pk?)]
       (try
         (let [parsed-rows (parse-rows rows)]
```

Comment:
just realized that we are realizing the whole "rows" here. 
I think it'd be nice if `driver/insert-into!` take a lazy seq and returns the number of items inserted. this way we don't have to have the whole seq realized before hand.

This is not a blocker a but a potential TODO